// @azure/core-sse/dist/commonjs/sse.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sse_exports = {};
__export(sse_exports, {
  createSseStream: () => createSseStream
});
module.exports = __toCommonJS(sse_exports);
var import_utils = require("./utils.js");
var ControlChars = /* @__PURE__ */ ((ControlChars2) => {
  ControlChars2[ControlChars2["NewLine"] = 10] = "NewLine";
  ControlChars2[ControlChars2["CarriageReturn"] = 13] = "CarriageReturn";
  ControlChars2[ControlChars2["Space"] = 32] = "Space";
  ControlChars2[ControlChars2["Colon"] = 58] = "Colon";
  return ControlChars2;
})(ControlChars || {});
function createSseStream(chunkStream) {
  const { cancel, iterable } = (0, import_utils.ensureAsyncIterable)(chunkStream);
  const asyncIter = toMessage(toLine(iterable));
  return (0, import_utils.createStream)(asyncIter, cancel);
}
function concatBuffer(a, b) {
  const res = new Uint8Array(a.length + b.length);
  res.set(a);
  res.set(b, a.length);
  return res;
}
function createMessage() {
  return {
    data: void 0,
    event: "",
    id: "",
    retry: void 0
  };
}
async function* toLine(chunkIter) {
  let buf;
  let bufIdx = 0;
  let fieldLen = -1;
  let discardTrailingNewline = false;
  for await (const chunk of chunkIter) {
    if (buf === void 0) {
      buf = chunk;
      bufIdx = 0;
      fieldLen = -1;
    } else {
      buf = concatBuffer(buf, chunk);
    }
    const bufLen = buf.length;
    let start = 0;
    while (bufIdx < bufLen) {
      if (discardTrailingNewline) {
        if (buf[bufIdx] === 10 /* NewLine */) {
          start = ++bufIdx;
        }
        discardTrailingNewline = false;
      }
      let end = -1;
      for (; bufIdx < bufLen && end === -1; ++bufIdx) {
        switch (buf[bufIdx]) {
          case 58 /* Colon */:
            if (fieldLen === -1) {
              fieldLen = bufIdx - start;
            }
            break;
          case 13 /* CarriageReturn */:
            discardTrailingNewline = true;
            end = bufIdx;
            break;
          case 10 /* NewLine */:
            end = bufIdx;
            break;
        }
      }
      if (end === -1) {
        break;
      }
      yield { line: buf.subarray(start, end), fieldLen };
      start = bufIdx;
      fieldLen = -1;
    }
    if (start === bufLen) {
      buf = void 0;
    } else if (start !== 0) {
      buf = buf.subarray(start);
      bufIdx -= start;
    }
  }
}
async function* toMessage(lineIter) {
  let message = createMessage();
  const decoder = new TextDecoder();
  for await (const { line, fieldLen } of lineIter) {
    if (line.length === 0 && message.data !== void 0) {
      yield message;
      message = createMessage();
    } else if (fieldLen > 0) {
      const field = decoder.decode(line.subarray(0, fieldLen));
      const valueOffset = fieldLen + (line[fieldLen + 1] === 32 /* Space */ ? 2 : 1);
      const value = decoder.decode(line.subarray(valueOffset));
      switch (field) {
        case "data":
          message.data = message.data ? message.data + "\n" + value : value;
          break;
        case "event":
          message.event = value;
          break;
        case "id":
          message.id = value;
          break;
        case "retry": {
          const retry = parseInt(value, 10);
          if (!isNaN(retry)) {
            message.retry = retry;
          }
          break;
        }
      }
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createSseStream
});
