// @azure/core-sse/dist/commonjs/utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  createStream: () => createStream,
  ensureAsyncIterable: () => ensureAsyncIterable
});
module.exports = __toCommonJS(utils_exports);
var import_types = require("./types.js");
function createStream(asyncIter, cancel) {
  const stream = iteratorToStream(asyncIter, cancel);
  return polyfillStream(stream, cancel);
}
function polyfillStream(stream, dispose) {
  makeAsyncIterable(stream);
  makeAsyncDisposable(stream, dispose);
  return stream;
}
function makeAsyncDisposable(webStream, dispose) {
  Symbol.asyncDispose ??= /* @__PURE__ */ Symbol("Symbol.asyncDispose");
  if (!webStream[Symbol.asyncDispose]) {
    webStream[Symbol.asyncDispose] = () => dispose();
  }
}
function makeAsyncIterable(webStream) {
  if (!webStream[Symbol.asyncIterator]) {
    webStream[Symbol.asyncIterator] = () => toAsyncIterable(webStream);
  }
  if (!webStream.values) {
    webStream.values = () => toAsyncIterable(webStream);
  }
}
function iteratorToStream(iterator, cancel) {
  return new ReadableStream({
    async pull(controller) {
      const { value, done } = await iterator.next();
      if (done) {
        controller.close();
      } else {
        controller.enqueue(value);
      }
    },
    cancel
  });
}
function ensureAsyncIterable(stream) {
  if (isReadableStream(stream)) {
    makeAsyncIterable(stream);
    return {
      cancel: () => stream.cancel(),
      iterable: stream
    };
  } else {
    return {
      cancel: async () => (0, import_types.cancelNodeStream)(stream),
      iterable: stream
    };
  }
}
function isReadableStream(body) {
  return Boolean(
    body && typeof body.getReader === "function" && typeof body.tee === "function"
  );
}
async function* toAsyncIterable(stream) {
  const reader = stream.getReader();
  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) {
        return;
      }
      yield value;
    }
  } finally {
    const cancelPromise = reader.cancel();
    reader.releaseLock();
    await cancelPromise;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createStream,
  ensureAsyncIterable
});
